addons/mass_mailing/models/mass_mailing.py

```diff
@@ -57,6 +57,54 @@ def _compute_contact_nbr(self):
         for mailing_list in self:
             mailing_list.contact_nbr = data.get(mailing_list.id, 0)
 
+    @api.multi
+    def copy(self, default=None):
+        recipients = self.env['mail.mass_mailing.contact'].search([('list_ids', 'in', self.ids), ('opt_out', '!=', True)])
+        mailing_list = super(MassMailingList, self).copy(default=default)
+        recipients.write({'list_ids': [(4, mailing_list.ids)]})
+        return mailing_list
+
+    @api.multi
+    def merge_massmail_list(self, src_mailing_lists, dst_mailing_lists):
+        self.env.cr.execute("""
+            SELECT DISTINCT contact.email, contact.id
+            FROM
+                mail_mass_mailing_contact contact,
+                mail_mass_mailing_contact_list_rel contact_list_rel,
+                mail_mass_mailing_list mass_mailing
+            WHERE contact.id=contact_list_rel.contact_id
+            AND contact.opt_out <> TRUE
+            AND mass_mailing.id=contact_list_rel.list_id
+            AND mass_mailing.id IN %s""",
+            (tuple(src_mailing_lists.ids), ))
+
+        contact_ids = [ x['id'] for x in  self.env.cr.dictfetchall()]
+        self.env['mail.mass_mailing.contact'].browse(contact_ids).write({'list_ids': [(4, contact_list.id) for contact_list in self]})
```

Comment:
Making a SQL query to browse the result after that and make the insertion with the ORM makes the SQL query useless. 

We should use INSERT INTO and do the things correctly in SQL